apps/condo/domains/miniapp/middlewares/launchB2BAppMiddleware/index.js

```diff
@@ -0,0 +1,132 @@
+const cookieParser = require('cookie-parser')
+const express = require('express')
+const { get } = require('lodash')
+
+const { getByCondition } = require('@open-condo/keystone/schema')
+
+const LaunchB2BAppHelpers = require('./helpers')
+
+
+class LaunchB2BAppMiddleware {
+    async prepareMiddleware () {
+        const helpers = LaunchB2BAppHelpers.getInstance()
+        const { config, jwksResponse, signLaunchParams } = helpers
+
+        const app = express()
+        app.use(cookieParser())
+
+        app.get('/api/miniapp/launch/jwks', (req, res) => {
+            // implementation from 'oidc-provider'
+            // analogue of redirect to '/oidc/jwks'
+            return res
+                .type('application/jwk-set+json; charset=utf-8')
+                .json(jwksResponse)
+        })
+
+        app.head('/api/miniapp/launch', async (req, res) => {
+
+            // 1) Check user is authorized
+            const authorizedUserId = get(req, 'user.id')
+            if (!authorizedUserId) return res.status(401).send('Unauthorized user')
+
+            // 2) Check req has appId, b2bApp, appUrl
+            const b2bAppId = get(req, 'query.id')
+            if (!b2bAppId) return res.status(400).send('No appId')
+
+            // 3) Check req from CRM
+            const referer = req.get('Referer')
+            if (!referer || config.whiteList.every((fn) => fn(b2bAppId) !== referer)) {
+                return res.status(403).send('Invalid source')
+            }
+
+            // 4) Check req has appUrl in b2bApp
+            const b2bApp = await getByCondition('B2BApp', {
+                id: b2bAppId,
+                deletedAt: null,
+            })
+            const appUrl = get(b2bApp, 'appUrl')
+            if (!appUrl) return res.status(400).send('Not found app url')
+
+            return res.redirect(appUrl)
+        })
+
+        app.get('/api/miniapp/launch', async (req, res) => {
+
+            // 1) Check user is authorized
+            const authorizedUserId = get(req, 'user.id')
+            if (!authorizedUserId) return res.status(401).send('Unauthorized user')
+
+            // 2) Check req has appId
+            const b2bAppId = get(req, 'query.id')
+            if (!b2bAppId) return res.status(400).send('No appId')
+
+            // 3) Check req from CRM
+            const referer = req.get('Referer')
+            if (!referer || config.whiteList.every((fn) => fn(b2bAppId) !== referer)) {
+                return res.status(403).send('Invalid source')
+            }
+
+            // 4) Check req has organizationLinkId
+            const employeeId = get(req, 'cookies.organizationLinkId')
+            if (!employeeId) return res.status(400).send('No organizationLinkId')
+
+            // 5) Check employee for authorized user
+            const employee = await getByCondition('OrganizationEmployee', {
```

Comment:
! SECURITY ISSUE ! you should check that the employee user is the same as the request user!

Reply:
It's better to check the user field explicitly